src/commands/ext-update.ts

```diff
@@ -60,148 +68,219 @@ export default new Command("ext:update <extensionInstanceId> [localDirectoryOrUr
+            `Extension instance '${clc.bold(instanceId)}' not found in project '${clc.bold(
+              projectId
+            )}'.`,
             {
               exit: 1,
             }
           );
         }
         throw err;
       }
-      const currentSpec: extensionsApi.ExtensionSpec = _.get(
+      const existingSpec: extensionsApi.ExtensionSpec = _.get(
         existingInstance,
         "config.source.spec"
       );
-      const currentParams = _.get(existingInstance, "config.params");
+      if (existingInstance.config.source.state === "DELETED") {
+        throw new FirebaseError(
+          `Instance '${clc.bold(
+            instanceId
+          )}' cannot be updated anymore because the underlying extension was unpublished from Firebase's registry of extensions. Going forward, you will only be able to re-configure or uninstall this instance.`
+        );
+      }
+      const existingParams = _.get(existingInstance, "config.params");
       const existingSource = _.get(existingInstance, "config.source.name");
 
-      let source;
-      let sourceName;
-      if (previews.extdev && directoryOrUrl) {
-        try {
-          source = await createSourceFromLocation(projectId, directoryOrUrl);
-          sourceName = source.name;
-        } catch (err) {
-          const invalidSourceErr = `Unable to update from the source \`${clc.bold(
-            directoryOrUrl
-          )}\`. To update this instance, you can either:\n
-          - Run \`${clc.bold(
-            "firebase ext:update " + instanceId
-          )}\` to update from the official source.\n
-          - Check your directory path or URL, then run \`${clc.bold(
-            "firebase ext:update " + instanceId + " <localDirectoryOrUrl>"
-          )}\` to update from a local directory or URL source.`;
-          throw new FirebaseError(invalidSourceErr);
+      // Infer updateSource if instance is from the registry
+      if (existingInstance.config.extensionRef && !updateSource) {
+        updateSource = `${existingInstance.config.extensionRef}@latest`;
+      } else if (existingInstance.config.extensionRef && semver.valid(updateSource)) {
+        updateSource = `${existingInstance.config.extensionRef}@${updateSource}`;
+      }
+
+      let newSourceName: string;
+      let published = false;
+
+      const existingSourceOrigin = await getExistingSourceOrigin(
+        projectId,
+        instanceId,
+        existingSpec.name,
+        existingSource
+      );
+      const newSourceOrigin = await getSourceOrigin(updateSource);
+
+      // We only allow the following types of updates.
+      let validUpdate = false;
+      if (existingSourceOrigin === SourceOrigin.OFFICIAL_EXTENSION) {
+        if (
+          [
+            SourceOrigin.LOCAL,
+            SourceOrigin.URL,
+            SourceOrigin.OFFICIAL_EXTENSION,
+            SourceOrigin.OFFICIAL_EXTENSION_VERSION,
+          ].indexOf(newSourceOrigin) > -1
```

Comment:
This seems to be the same list (and check) as later in this file. Consider moving the list to a constant above; and consider doing something like this (rather than .indexOf)

```typescript
allowedSourceOrigins.includes(newSourceOrigin)
```

Reply:
It's actually a slightly different list. I would prefer to continue being explicit here so that someone can analyze the behavior more easily. There's more context shift if they have to look elsewhere, and in this particular case, I think having all the context in one place is nice because this part of the code is quite complicated (and should be refactored/simplified in Q1 when we migrate official extensions on the Registry).